sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -37,6 +37,12 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
     // SubQueries are only needed for analysis and can be removed before execution.
     Batch("Remove SubQueries", FixedPoint(100),
       EliminateSubQueries) ::
+    // - Do the first call of CombineUnions before starting the major Optimizer rules,
+    //   since they could add/move extra operators between two adjacent Union operators.
+    // - Call CombineUnions again in Batch("Operator Optimizations"),
+    //   since the other rules might make two separate Unions operators adjacent.
+    Batch("Union", FixedPoint(100),
+      CombineUnions) ::
```

Comment:
I will also add performance issue in the comment.

@cloud-fan https://github.com/apache/spark/pull/10451 This PR will add extra `Limit` through Union. I guess, if we do multiple `Union All` with one `Limit` in SQL, the plan could be changed like 
```
Limit 
Union 
Limit 
Union 
Limit
...
```